cpp/include/cudf/detail/aggregation/aggregation.hpp

```diff
@@ -22,83 +22,146 @@
+    : derived_aggregation{QUANTILE}, _quantiles{q}, _interpolation{i}
   {
   }
   std::vector<double> _quantiles;              ///< Desired quantile(s)
   experimental::interpolation _interpolation;  ///< Desired interpolation
 
+ protected:
+  friend class derived_aggregation<quantile_aggregation>;
+
   bool operator==(quantile_aggregation const& other) const
   {
-    return aggregation::operator==(other) and _interpolation == other._interpolation and
+    return _interpolation == other._interpolation and
            std::equal(_quantiles.begin(), _quantiles.end(), other._quantiles.begin());
   }
+
+  size_t hash_impl() const
+  {
+    return std::hash<int>{}(static_cast<int>(_interpolation)) ^
+           std::accumulate(_quantiles.cbegin(), _quantiles.cend(), 0, [](size_t a, double b) {
+             return a ^ std::hash<double>{}(b);
+           });
+  }
 };
 
 /**
  * @brief Derived class for specifying a standard deviation/variance aggregation
  */
-struct std_var_aggregation : aggregation {
-  std_var_aggregation(aggregation::Kind k, size_type ddof) : aggregation{k}, _ddof{ddof} {}
+struct std_var_aggregation final : derived_aggregation<std_var_aggregation> {
+  std_var_aggregation(aggregation::Kind k, size_type ddof) : derived_aggregation{k}, _ddof{ddof} {}
   size_type _ddof;  ///< Delta degrees of freedom
 
-  bool operator==(std_var_aggregation const& other) const
-  {
-    return aggregation::operator==(other) and _ddof == other._ddof;
-  }
+ protected:
+  friend class derived_aggregation<std_var_aggregation>;
+
+  bool operator==(std_var_aggregation const& other) const { return _ddof == other._ddof; }
+
+  size_t hash_impl() const { return std::hash<size_type>{}(_ddof); }
 };
 
 /**
  * @brief Derived class for specifying a nunique aggregation
  */
-struct nunique_aggregation : aggregation {
+struct nunique_aggregation final : derived_aggregation<nunique_aggregation> {
   nunique_aggregation(aggregation::Kind k, include_nulls _include_nulls)
-    : aggregation{k}, _include_nulls{_include_nulls}
+    : derived_aggregation{k}, _include_nulls{_include_nulls}
   {
   }
   include_nulls _include_nulls;  ///< include or exclude nulls
 
+ protected:
+  friend class derived_aggregation<nunique_aggregation>;
+
   bool operator==(nunique_aggregation const& other) const
   {
-    return aggregation::operator==(other) and _include_nulls == other._include_nulls;
+    return _include_nulls == other._include_nulls;
   }
+
+  size_t hash_impl() const { return std::hash<int>{}(static_cast<int>(_include_nulls)); }
 };
 
 /**
  * @brief Derived class for specifying a nth element aggregation
  */
-struct nth_element_aggregation : aggregation {
+struct nth_element_aggregation final : derived_aggregation<nth_element_aggregation> {
   nth_element_aggregation(aggregation::Kind k, size_type n, include_nulls _include_nulls)
-    : aggregation{k}, n{n}, _include_nulls{_include_nulls}
+    : derived_aggregation{k}, n{n}, _include_nulls{_include_nulls}
```

Comment:
Not sure how I feel about same name parameters and variables. Do we have a policy on this in RAPIDS? @jrhemstad

Reply:
Oh, good point, didn't notice that. @harrism wasn't a fan of `include_nulls` vs `bool` on another PR since it results in this awkward `include_nulls _include_nulls` signature.

Reply:
That was addressed in this commit. I suppose it would be an improvement here 🤷‍♂️ 
https://github.com/rapidsai/cudf/pull/4739/commits/98cb3b2cf382babbc56332e3b652ad0b5ef0a3a7

Reply:
Added `_` prefixes to the relevant members here when merging the `null_policy` changes.